### PR TITLE
fix: crash when sequence packing is enabled for gemma 1b.

### LINKED ISSUE #790: sequence_packing does not work for gemma models
**Describe the bug**

sequence_packing=T does not work for Gemma model. The reason I notice this is because sequence_packing=T seems to be the default for llama, but when i change the model to Gemma with the same config, it failed with Cuda error below. 

If this is the intended,  please add config validation and better error message.

**Steps/Code to reproduce bug**
- Gemma model worked for dynamic_batching=T and sequence_packing=F (default)
`uv run python examples/run_grpo_math.py --config examples/configs/recipes/llm/grpo-gemma3-1b-it-1n8g-fsdp2tp1.yaml` 
- but it failed with dynamic_batching=F and sequence_packing=T with cuda error when computing logprob
`uv run python examples/run_grpo_math.py --config examples/configs/recipes/llm/grpo-gemma3-1b-it-1n8g-fsdp2tp1.yaml policy.dynamic_batching.enabled=False policy.sequence_packing.enabled=True ++policy.megatron_cfg.enabled=false` 
```
▶ Computing logprobs...
2025-07-29 19:40:28
(DTensorPolicyWorker[rank=0] pid=1925039) /pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:144: operator(): block: [622,0,0], thread: [0,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
```

### PR DESCRIPTION
**Pull request by joyang-nv**
# What does this PR do ?

Fix crash when get_logprobs for dtensor with gemma model...
The reason why we don't have this bug on other models like llama or qwen is because we passed all ones attention mask and inside modeling script will change attention mask to pass to following on operators to None.
But for model like gemma won't take such approach. And attention mask shape mismatch with input_ids results in crash (Only in flash attention2 now).

# Issues
sequence_packing does not work for gemma models. closes #790

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
Verify with command to compare with dynamic_batching enabled case (pink):
```
NRL_FORCE_REBUILD_VENVS=1 uv run python examples/run_grpo_math.py  \
    --config examples/configs/recipes/llm/grpo-gemma3-1b-it-1n8g-fsdp2tp1.yaml \
    cluster.gpus_per_node=2 \
    grpo.max_num_steps=25 \
    policy.dynamic_batching.enabled=False \
    policy.sequence_packing.enabled=True \
    ++policy.dtensor_cfg._v2=0 \
    ++policy.sequence_packing.train_mb_tokens=2048 \
    ++policy.sequence_packing.logprob_mb_tokens=2048 \
    ++policy.sequence_packing.algorithm="first_fit_decreasing" \
    ++policy.sequence_packing.sequence_length_round=64 \
    ++policy.megatron_cfg.enabled=false \
    logger.wandb_enabled=1 \
    ++logger.wandb.project=${WANDB_PROJECT} \
    ++logger.wandb.entity=${WANDB_ENTITY} \
    grpo.num_prompts_per_step=8 policy.train_global_batch_size=128 policy.train_micro_batch_size=16

```
Graph with commit c2ebb8c7886fcbcd2f91df2ad994f372b7ace49c.
image

image

image

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed incorrect handling of padding for right‑padded (non-packed) inputs so padded tokens no longer affect token log probability calculations.
  * Sequence‑packing behavior remains unchanged.

* **Refactor**
  * Mask handling unified to boolean and split into forward‑pass vs post‑processing masks for more consistent, reliable behavior.
  * No changes to public APIs.

### CODE REVIEW

**Comment by SahilJain314** [comment]
I see that the training function has an attention mask of None when Sequence Packing is enabled. Do you know why we have to do it differently for logprobs?

**Comment by joyang-nv** [comment]
Rebased!

**Comment by terrykong** [comment]
@coderabbitai generate sequence diagram

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 1]
```diff
@@ -995,14 +995,16 @@ def get_logprobs(
...
 
-                with torch.autocast(device_type="cuda", dtype=self.dtype):
-                    # DTensor requires the casual attention kernel to hit,
-                    # yet our attention mask above is not always all 1s
-                    # this is fine because we mask with the actual attention mask
-                    # later, but for input it has to be all 1s
-                    attention_mask_input_all_ones = torch.ones(
-                        (batch_size, seq_len), dtype=torch.long, device=input_ids.device
-                    )
+                    with torch.autocast(device_type="cuda", dtype=self.dtype):
```
Comment (terrykong) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: does this autocast do anything?

**Review by terrykong — COMMENTED** [review]
@joyang-nv  perhaps i'm missing something, but aside from the indentation change, it looks like the logic is unchanged, but the variable names have changed. why exactly does this fix gemma?

File: nemo_rl/models/policy/dtensor_policy_worker.py (hunk 1 above)
Comment (joyang-nv) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: > why exactly does this fix gemma?
Terry, the crash was caused by improper `attention_mask` argument when sequence pack is enabled. The root cause is tricky:

1.  Why it doesn't crash on other models?
    It is because models like qwen will treat all ones attention_mask as None when model forward is called and inside [this function](https://github.com/huggingface/transformers/blob/5af248b3e3bf854e884c53f05b17ed1e0ca2ed24/src/transformers/masking_utils.py#L744). So the actual `attention_mask` arg passed to attention module is also None. But this is not the case for gemma.
3. Why crashing?
When sequence packing is enabled, all zeros attention mask is not correct and passed to attention module. This results crash in kernel.

Comment (joyang-nv) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: BTW, some observations during fix:
1. I can't reproduce this on H100.
2. I can't reproduce on policy v2, but I did the same fix to make them synced.
3. I can reproduce this on RTA A6000.

Comment (joyang-nv) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: For `autocast`, I will check whether we need it.

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 2]
```diff
@@ -672,7 +672,7 @@ def train(
 
                             attention_mask = torch.ones(
                                 (batch_size, seq_len),
-                                dtype=torch.long,
+                                dtype=torch.bool,
```
Comment (yuki-97) [inline nemo_rl/models/policy/dtensor_policy_worker.py:675]: is there any benefit for us to change the dtype from long to bool?

in my impression, attention_mask is a long dtype for the convenience of calculation. it need to be converted to long/float if it's a bool when calculating. please correct me if i'm wrong.

Comment (joyang-nv) [reply nemo_rl/models/policy/dtensor_policy_worker.py:675]: Yes, it is. But you could leave torch to do implicit conversion for you. Torch.bool saves memory.

Comment (joyang-nv) [reply nemo_rl/models/policy/dtensor_policy_worker.py:675]: I don't think we should use long (int64) for bit mask tensor. Bool type matches more intuitive for variable name "mask". Even it has introduced conversion, let torch does it for only when necessary.

**Review by terrykong — APPROVED** [review]
